### PR TITLE
Integration of Square Payment Gateway

### PR DESCRIPTION
**Pull request by gustavop-dev**
# Integration of Square Payment Gateway

## Description

This PR implements the integration of the Square payment gateway into our system using Django for the backend and Vue.js for the frontend. The integration allows users to make payments directly through our application using Square's Payments API.

### Frontend (Vue.js)

**1. Payment Form Component:**
   - Created a new `PaymentForm.vue` component to handle the credit card tokenization using Square's SDK.
   - Added logic to manage and send the payment token to the backend.

**2. Create Sale Function:**
   - Updated the `createSale` function in the Vue store to send the necessary data to the backend and handle the response appropriately.
   - Example usage of `createSale`:
     ```javascript
     async createSale(form) {
       try {
         const response = await create_request('create-sale/', {
           email: form.email,
           address: form.address,
           city: form.city,
           state: form.state,
           postal_code: form.postalCode,
           sold_products: form.soldProducts,
           total: form.total,
           sourceId: form.sourceId,
           locationId: form.locationId,
           idempotencyKey: Math.random().toString(36).substring(2),
         });

         return response.status;
       } catch (error) {
         console.error('Error creating sale:', error);
         throw error;
       }
     }
     ```

**3. Checkout View:**
   - Modified the `handleSubmit` function to handle the entire payment process, including tokenization, sale creation, and response handling.
   - Example:
     ```javascript
     const handleSubmit = async () => {
       try {
         const token = await paymentFormRef.value.handlePaymentMethodSubmission();
         form.sourceId = token;
         form.soldProducts = extractProductInfo(cartProducts.value);
         const status = await productStore.createSale(form);

         if (status === 201) {
           Swal.fire({
             title: "The payment has been processed successfully",
             icon: "success"
           });

           localStorage.removeItem("cartProducts");
           productStore.cartProducts = [];
           router.push({ name: "home" });
         } else {
           Swal.fire({
             title: "Payment failed",
             text: "An error occurred while processing your payment",
             icon: "error"
           });
         }
       } catch (error) {
         console.error('Error during payment processing:', error);
         Swal.fire({
           title: "Payment failed",
           text: error.message,
           icon: "error"
         });
       }
     };
     ```

### Backend (Django)

**1. Create Sale Endpoint:**
   - Created a new API endpoint in Django to handle the payment processing and sale creation.
   - Configured the Square client and integrated the payment processing logic.
   - Example:
     ```python
     from rest_framework.decorators import api_view
     from rest_framework.response import Response
     from rest_framework import status
     from square.client import Client
     from candle_app.serializers import SaleSerializer
     import uuid

     @api_view(['POST'])
     def create_sale(request):
         if request.method == 'POST':
             data = request.data
             source_id = data.get('sourceId')
             location_id = data.get('locationId')
             idempotency_key = data.get('idempotencyKey')
             total = data.get('total')

             if not total:
                 return Response({"error": "Total amount is required"}, status=status.HTTP_400_BAD_REQUEST)

             amount_in_cents = int(float(total) * 100)

             client = Client(
                 access_token='YOUR_SQUARE_ACCESS_TOKEN',
                 environment='sandbox'
             )
             payments_api = client.payments

             body = {
                 "source_id": source_id,
                 "location_id": location_id,
                 "idempotency_key": idempotency_key,
                 "amount_money": {
                     "amount": amount_in_cents,
                     "currency": "USD"
                 }
             }

             try:
                 result = payments_api.create_payment(body)
                 if result.is_success():
                     serializer = SaleSerializer(data=data)
                     if serializer.is_valid():
                         serializer.save()
                         return Response(serializer.data, status=status.HTTP_201_CREATED)
                     return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
                 else:
                     return Response(result.errors, status=status.HTTP_400_BAD_REQUEST)
             except Exception as e:
                 return Response({"error": str(e)}, status=status.HTTP_500_INTERNAL_SERVER_ERROR)
     ```

## How to Test

**1. Frontend:**
   - Navigate to the checkout page and fill out the payment form.
   - Submit the form and verify that the payment is processed and a success message is displayed.
   - Check that the cart is cleared and you are redirected to the home page.

**2. Backend:**
   - Verify that the `create-sale` endpoint is receiving the correct data and processing payments successfully.
   - Ensure that sales are created in the database and all relevant data is stored correctly.

## Notes

- Make sure to replace `YOUR_SQUARE_ACCESS_TOKEN` with your actual Square access token.
- This integration is currently set up for the Square sandbox environment. Change the environment to 'production' when ready to go live.
- For 'production' It's neccesary change:

- `YOUR_SQUARE_ACCESS_TOKEN` 
- `YOUR_SQUARE_APPLICATION_ID` 
- `YOUR_SQUARE_LOCATION_ID` 
- `<script src="https://sandbox.web.squarecdn.com/v1/square.js"></script>` to `<script src="https://web.squarecdn.com/v1/square.js"></script>`

### CODE REVIEW

File: frontend/src/components/alerts/AlertShoppingCart.vue [hunk 1]
```diff
@@ -0,0 +1,90 @@
+<template>
+    <div v-if="visible" class="w-screen h-14 bg-primary_p flex justify-between items-center px-16 text-white font-regular text-md shadow-2xl">
+        <div>
+            <p class="test-addShoppingCart-text">{{ $t('addShoppingCart').text }}</p>
+        </div>
+        <div class="flex gap-6">
+            <a @click="shoppingCartToggle = true" class="cursor-pointer underline test-addShoppingCart-seeCart">{{ $t('addShoppingCart').seeCart }}</a>
```
Comment (carlos18bp) [inline frontend/src/components/alerts/AlertShoppingCart.vue:7]: Change to:
```
<a @click="shoppingCartToggle = true" 
   class="cursor-pointer underline test-addShoppingCart-seeCart">
   {{ $t('addShoppingCart').seeCart }}
</a>
```

File: frontend/src/components/alerts/AlertShoppingCart.vue [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+            <XMarkIcon @click="closeModal()" class="w-6 h-6 cursor-pointer"></XMarkIcon>
+        </div>
+    </div>
+
+    <!-- Shopping cart component -->
+    <div v-if="shoppingCartToggle" class="fixed z-30 w-full h-screen top-0">
+        <ShoppingCart :visible="shoppingCartToggle" @update:visible="shoppingCartToggle = $event"></ShoppingCart>
```
Comment (carlos18bp) [inline frontend/src/components/alerts/AlertShoppingCart.vue:14]: Change to:
```
<ShoppingCart :visible="shoppingCartToggle" 
   @update:visible="shoppingCartToggle = $event">
</ShoppingCart>
```

File: frontend/src/components/animations/Loading.vue [hunk 3]
```diff
@@ -0,0 +1,31 @@
+<template>
+  <div>
+    <Vue3Lottie :animationData="animationData" :height="height" :width="width" :loop="loop" :autoplay="autoplay" />
```
Comment (carlos18bp) [inline frontend/src/components/animations/Loading.vue:3]: Change to:
```
 <Vue3Lottie :animationData="animationData" 
   :height="height" 
   :width="width" 
   :loop="loop" 
   :autoplay="autoplay" />
```

File: frontend/src/components/animations/ScreenLoading.vue [hunk 4]
```diff
@@ -0,0 +1,15 @@
+<template>
+    <div class="fixed inset-0 flex justify-end z-50">
+        <div 
+            class="absolute inset-0 bg-gray-500 bg-opacity-40 backdrop-blur-md">
+        </div>
+        <div class="w-screen h-screen flex justify-center items-center z-60">
+            <Vue3Lottie :animationData="animationData" :height="600" :width="600" :loop="true" :autoplay="true"></Vue3Lottie>
```
Comment (carlos18bp) [inline frontend/src/components/animations/ScreenLoading.vue:7]: Change to:
```
<Vue3Lottie :animationData="animationData" 
   :height="600" 
   :width="600" 
   :loop="true" 
   :autoplay="true">
</Vue3Lottie>
```

File: frontend/src/components/payments/PaymentForm.vue [hunk 5]
```diff
@@ -0,0 +1,82 @@
+<template>
+    <form @submit.prevent="handlePaymentMethodSubmission">
+      <div v-if="loading">
+        <LoadingAnimation :height="300" :width="300" :loop="true" :autoplay="true"></LoadingAnimation>
```
Comment (carlos18bp) [inline frontend/src/components/payments/PaymentForm.vue:4]: Change to:
```
<LoadingAnimation :height="300" 
   :width="300" 
   :loop="true" 
   :autoplay="true">
</LoadingAnimation>
```

File: frontend/src/components/payments/PaymentForm.vue [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+        throw new Error("Square.js failed to load properly");
+    }
+    const payments = Square.payments(appId, locationId);
+    try {
+        card = await payments.card();
+        await card.attach("#card-container");
+    } catch (e) {
+        console.error("Initializing Card failed", e);
+        return;
+    }
```
Comment (carlos18bp) [inline frontend/src/components/payments/PaymentForm.vue:37]: Add one tab here

File: frontend/src/components/payments/PaymentForm.vue [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+     */
+    const initializePaymentForm = async () => {
+    if (!Square) {
+        throw new Error("Square.js failed to load properly");
+    }
+    const payments = Square.payments(appId, locationId);
+    try {
+        card = await payments.card();
+        await card.attach("#card-container");
+    } catch (e) {
```
Comment (carlos18bp) [inline frontend/src/components/payments/PaymentForm.vue:34]: Change `e` to `error`

File: frontend/src/components/payments/PaymentForm.vue [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+    const tokenResult = await paymentMethod.tokenize();
+    if (tokenResult.status === "OK") {
+        return tokenResult.token;
+    } else {
+        let errorMessage = `Tokenization failed with status: ${tokenResult.status}`;
+        if (tokenResult.errors) {
+        errorMessage += ` and errors: ${JSON.stringify(tokenResult.errors)}`;
+        }
+        throw new Error(errorMessage);
+    }
```
Comment (carlos18bp) [inline frontend/src/components/payments/PaymentForm.vue:56]: Add one tab here

File: frontend/src/components/payments/PaymentForm.vue [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+     * Tokenizes the payment method and emits the event with the token.
+     */
+    const handlePaymentMethodSubmission = async () => {
+        const token = await tokenize(card);
+        return token
+    };
+
+    // Expose the handlePaymentMethodSubmission function to be called externally
+    defineExpose({
+    handlePaymentMethodSubmission,
```
Comment (carlos18bp) [inline frontend/src/components/payments/PaymentForm.vue:70]: Add one tab here

File: frontend/src/components/payments/PaymentForm.vue [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+    });
+
+    /**
+     * Vue hook that runs when the component is mounted.
+     * Initializes the payment form and updates the loading state.
+     */
+    onMounted(async () => {
+    loading.value = true;
+    await initializePaymentForm();
+    loading.value = false;
```
Comment (carlos18bp) [inline frontend/src/components/payments/PaymentForm.vue:80]: Add one tab here

File: frontend/src/views/product/Detail.vue [hunk 11]
```diff
@@ -293,6 +293,11 @@
 
     <!-- Footer Component -->
     <Footer></Footer>
+
+     <!-- Alert component -->
+     <div v-if="showAlertShoppingCart" class="fixed z-30 top-0">
+        <AlertShoppingCart :visible="showAlertShoppingCart" @update:visible="showAlertShoppingCart = $event"></AlertShoppingCart>
```
Comment (carlos18bp) [inline frontend/src/views/product/Detail.vue:299]: Change to:
```
<AlertShoppingCart :visible="showAlertShoppingCart" 
   @update:visible="showAlertShoppingCart = $event">
</AlertShoppingCart>
```

**Review by carlos18bp — CHANGES_REQUESTED** [review]

File: frontend/src/components/alerts/AlertShoppingCart.vue (hunk 1 above)
Comment (gustavop-dev) [reply frontend/src/components/alerts/AlertShoppingCart.vue:7]: Done!

File: frontend/src/components/alerts/AlertShoppingCart.vue (hunk 2 above)
Comment (gustavop-dev) [reply frontend/src/components/alerts/AlertShoppingCart.vue:14]: Done!

File: frontend/src/components/animations/Loading.vue (hunk 3 above)
Comment (gustavop-dev) [reply frontend/src/components/animations/Loading.vue:3]: Done!

File: frontend/src/components/animations/ScreenLoading.vue (hunk 4 above)
Comment (gustavop-dev) [reply frontend/src/components/animations/ScreenLoading.vue:7]: Done!

File: frontend/src/components/payments/PaymentForm.vue (hunk 5 above)
Comment (gustavop-dev) [reply frontend/src/components/payments/PaymentForm.vue:4]: Done!

File: frontend/src/components/payments/PaymentForm.vue (hunk 6 above)
Comment (gustavop-dev) [reply frontend/src/components/payments/PaymentForm.vue:37]: Done!

File: frontend/src/components/payments/PaymentForm.vue (hunk 7 above)
Comment (gustavop-dev) [reply frontend/src/components/payments/PaymentForm.vue:34]: Done!

File: frontend/src/components/payments/PaymentForm.vue (hunk 8 above)
Comment (gustavop-dev) [reply frontend/src/components/payments/PaymentForm.vue:56]: Done!

File: frontend/src/components/payments/PaymentForm.vue (hunk 9 above)
Comment (gustavop-dev) [reply frontend/src/components/payments/PaymentForm.vue:70]: Done!

File: frontend/src/components/payments/PaymentForm.vue (hunk 10 above)
Comment (gustavop-dev) [reply frontend/src/components/payments/PaymentForm.vue:80]: Done!

File: frontend/src/views/product/Detail.vue (hunk 11 above)
Comment (gustavop-dev) [reply frontend/src/views/product/Detail.vue:299]: Done!

File: frontend/src/components/alerts/AlertShoppingCart.vue [hunk 12]
```diff
@@ -0,0 +1,94 @@
+<template>
+    <div v-if="visible" class="w-screen h-14 bg-primary_p flex justify-between items-center px-16 text-white font-regular text-md shadow-2xl">
+        <div>
+            <p class="test-addShoppingCart-text">{{ $t('addShoppingCart').text }}</p>
+        </div>
+        <div class="flex gap-6">
+            <a @click="shoppingCartToggle = true" class="cursor-pointer underline test-addShoppingCart-seeCart">
```
Comment (carlos18bp) [inline frontend/src/components/alerts/AlertShoppingCart.vue:7]: Change to: 
```
<a @click="shoppingCartToggle = true" 
   class="cursor-pointer underline test-addShoppingCart-seeCart">
```

**Review by carlos18bp — APPROVED** [review]

File: frontend/src/components/alerts/AlertShoppingCart.vue (hunk 12 above)
Comment (gustavop-dev) [reply frontend/src/components/alerts/AlertShoppingCart.vue]: Done!

**Pull request merged**